### ISSUE TITLE
Missing grad_fn when passing a simple tensor through the reformer module.

### ISSUE THREAD

**Issue by pabloppp** [lucidrains/reformer-pytorch#29]
If I try to pass a simple tensor, that does not require grad, through a Reformer, it won't allow me to do backpropagation.

```python
x = torch.randn(batch_size, 256, 512).cuda()  
pred = reformer(x)
loss = criterion(pred, torch.ones_like(pred))
optimizer.zero_grad()
loss.backward()
optimizer.step()
```

<img width="1007" alt="Captura de pantalla 2020-02-10 a las 12 35 44" src="https://user-images.githubusercontent.com/4237014/74146649-dd516a00-4c01-11ea-909e-97663a91e2a5.png">

If I do `x.requires_grad = True` just before passing it to the reformer model it works though.

I am not sure why this happens, but makes me think that the model is not being optimized at all when training :/

**Comment by lucidrains** [comment]
Ohh, I understand why, it's because the reversible networks executes the self attention and feed forward network with a `no_grad()` https://github.com/RobinBruegger/RevTorch/blob/master/revtorch/revtorch.py#L47 , and does the gradient computation manually on the backward pass. When using a Reformer, it believes there is nothing to optimize. You can trick it into working by instantiating `x` as `torch.randn(batch_size, 256, 512, requires_grad=True).cuda()`

**Comment by lucidrains** [comment]
I'll leave this issue open, because I'm wondering if any Pytorch experts out there know any better way to make it work by default, in the case that the entire network consists only of the Reformer
